aten/src/ATen/native/xnnpack/Pooling.cpp

```diff
@@ -0,0 +1,321 @@
+#ifdef USE_XNNPACK
+
+#include <ATen/native/Pool.h>
+#include <ATen/native/utils/ParamUtils.h>
+#include <ATen/native/xnnpack/Common.h>
+#include <ATen/native/xnnpack/Factory.h>
+
+namespace at {
+namespace native {
+namespace xnnpack {
+namespace internal {
+namespace max_pool2d {
+
+struct Context final {
+  Operator max_pool_op;
+
+  struct Output final {
+    struct Spatial final {
+      int64_t kernel;
+      int64_t padding;
+      int64_t stride;
+      int64_t dilation;
+      bool ceil_mode;
+    };
+
+    int64_t channels;
+    Spatial height, width;
+  } output;
+
+  static constexpr float kMin = -std::numeric_limits<float>::infinity();
+  static constexpr float kMax = std::numeric_limits<float>::infinity();
+};
+
+namespace {
+
+// Supports NHWC and NCHW FP32 max pooling with any
+//  - kernel size
+//  - padding
+//  - stride
+//  - dilation
+
+bool available(
+    const int64_t input_channels,
+    const int64_t output_channels,
+    const IntArrayRef kernel,
+    const IntArrayRef padding,
+    const IntArrayRef stride,
+    const IntArrayRef dilation,
+    const bool /* ceil_mode */,
+    const float output_min,
+    const float output_max) {
+         // XNNPACK
+  return xnnpack::internal::available() &&
+         // Input
+         (input_channels > 0) &&
+         // Input == Output
+         (input_channels == output_channels) &&
+         // Kernel
+         ((1 <= kernel.size()) && (kernel.size() <= 2)) &&
+         (kernel[Layout::Parameter::height] > 0) &&
+         (kernel[Layout::Parameter::width] > 0) &&
```

Comment:
how does this work when kernel.size()==1? Layout::Parameter::width is 1, and you can't have kernel[1]. Ah, I see, you are expanding `kernel` in `create`, but then check for kernel.size() should be different.

Reply:
I'm sorry I didn't get your point about kernel.size() being different, can you elaborate?  Yes, I'm allowing 1 and 2 but expanding to 2 prior to use.

Reply:
If you require 2, then you should not allow 1, and looks like you require 2, because otherwise kernel[1] will segfault.

Reply:
Aaah ... how could I have missed that?!  :/  Thanks.